app/guided.py

```diff
@@ -0,0 +1,413 @@
+        last_action_absolute_score = absolute_dist_table[self.last_action.ability_id]
+        effective_dist_table[self.last_action.ability_id] = self._apply_effective_distance_penalty(
+            last_action_effective_score, last_action_absolute_score, executable_link_max_val)
+
+        for ability_id, score in effective_dist_table.items():
+            effective_dist_table[ability_id] = score + float(absolute_dist_table[ability_id] - score) / \
+                                               self.half_life_gain
+        return effective_dist_table
+
+    def _apply_effective_distance_penalty(self, effective_score, absolute_score, max_value):
+        """
+        Calculates the new effective distance for the last action performed.
+        """
+        if self.last_action.ability_id in self.goal_actions:
+            new_distance = max_value - 1 + float(effective_score + 1 - max_value) / self.goal_action_decay  # Action was a goal action, decentivize this action to avoid local optimia
+        elif max_value > abs(absolute_score):
+            new_distance = absolute_score   # we're now closer to our goal, reset the effective distance of the last action
+        else:
+            new_distance = float(effective_score) / self.half_life_penalty
+        
+        return new_distance
+
+    async def _get_goal_links(self, absolute_distance_table, agent=None):
+        """
+        Produces all links for abilities that are present in the absolute distance table.
+        """
+        links = await self.planning_svc.get_links(self.operation, agent=agent)
+        return [link for link in links if link.ability.ability_id in absolute_distance_table]
+
+    async def _task_agents(self, links, link_distance_table, abilities, agent=None):
+        """
+        Chooses next actions for agent provided. If no agent is provided, then the method will choose and return
+        next actions for all agents in the operation.
+        """
+        tasked_link_ids = []
+        agents = [agent] if agent else self.operation.agents
+        for operation_agent in agents:
+            agent_links = [link for link in links if link.paw == operation_agent.paw]
+            if not len(agent_links):
+                continue
+            tasked_link = await self._task_best_action(agent_links, link_distance_table, abilities)
+            tasked_link_ids.append(tasked_link.id)
+
+        return tasked_link_ids
+
+    async def _task_best_action(self, agent_links, link_distance_table, abilities):
+        """
+        Produces the best available link based on the current distance table. The method also applies links
+        that support the chosen link, but that do not actively work towards the goal.
+        """
+        async def _heapsort(links):
+            sorted_links = []
+            counter = itertools.count()
+            for _, (dist, link) in links.items():
+                entry = [-dist, -link.score, next(counter), link]
```

Comment:
what value is the heap being sorted on?

Reply:
Primarily sorted on distance from link to a goal, with ties broken on the score of the link and an autoincrementing counter. I am renaming some variables and adding a comment explaining this.